Refactored licensing module

akallabeth opened the pull request:
* Make the whole module opaque for easier testing

hardening commented on libfreerdp/core/license.c:
```diff
@@ -542,14 +910,115 @@ int license_recv(rdpLicense* license, wStream* s)
+				goto fail;
+
+			// TODO: validate challenge response
+			if (FALSE)
```
Shall we merge the PR and you'll work that in a second step ?

hardening commented on libfreerdp/core/license.c:
```diff
@@ -1703,7 +2646,173 @@ void license_free(rdpLicense* license)
+	if (!s)
+		return FALSE;
+	else
+	{
```
we can probably remove the `else` (not that we have upgraded C standard, declaring `r` right after should be ok)

akallabeth replied:
true, but I think the scope makes it more readable, but can change that.

hardening commented:
Overall looks good to me.
There's some TODOs in the code, and for some of them they seem like rest of coding work, could check these ?

MartinHaimberger commented on libfreerdp/core/license.c:
```diff
@@ -1492,56 +2374,38 @@ BOOL license_answer_license_request(rdpLicense* license)
-	BYTE mac_data[16];
 	BOOL status;
 
+	WINPR_ASSERT(license);
```
already checked in license_send_stream_init

akallabeth commented on libfreerdp/core/connection.c:
```diff
@@ -1053,23 +1053,40 @@ BOOL rdp_client_connect_auto_detect(rdpRdp* rdp, wStream* s)
 
-	if (status < 0)
-		return status;
+	if (!rdp_read_header(rdp, s, &length, &channelId))
```
not required, the pointers are not dereferenced and the function called do the assertions

akallabeth commented:
@MartinHaimberger ok, added/removed assertions, only kept the ones that are required as the pointers are dereferenced and removed the ones that are only passed on

MartinHaimberger reviewed (approved):
still one not needed ASSERT but otherwise it looks good
